gzip/gzip.go

```diff
@@ -0,0 +1,83 @@
+package gzip
+
+import (
+	"bytes"
+	"compress/gzip"
+	"io/ioutil"
+
+	kafka "github.com/segmentio/kafka-go"
+)
+
+func init() {
+	kafka.RegisterCompressionCodec(func() kafka.CompressionCodec {
+		return NewCompressionCodec()
+	})
+}
+
+type CompressionCodec struct {
+	CompressionLevel int
+}
+
+func NewCompressionCodec() CompressionCodec {
+	return NewCompressionCodecWith(kafka.DefaultCompressionLevel)
+}
+
+func NewCompressionCodecWith(level int) CompressionCodec {
+	return CompressionCodec{
+		CompressionLevel: level,
+	}
+}
+
+// Code implements the kafka.CompressionCodec interface.
+func (c CompressionCodec) Code() int8 {
+	return 1
+}
+
+// Encode implements the kafka.CompressionCodec interface.
+func (c CompressionCodec) Encode(dst, src []byte) (int, error) {
+	var buf bytes.Buffer
+	writer := gzip.NewWriter(&buf)
+	_, err := writer.Write(src)
+	if err != nil {
+		return 0, err
+	}
+	err = writer.Close()
+	if err != nil {
+		return 0, err
+	}
+
+	n, err := buf.WriteTo(&buffer{
+		data: dst,
+	})
+	return int(n), err
+}
+
+// Decode implements the kafka.CompressionCodec interface.
+func (c CompressionCodec) Decode(dst, src []byte) (int, error) {
+	reader, err := gzip.NewReader(bytes.NewReader(src))
```

Comment:
Same comment as for the writer, reusing the reader and calling it's `Reset` method to configure its input would be much more efficient here.